pkg/ipcache/ipcache.go

```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+					scopedLog.Info("Upserting endpoint IP shadowing CIDR")
+					oldIdentity = &cidrIdentity.ID
+				} else {
+					// The endpoint IP and the CIDR are associated with the
+					// same identity. Nothing changes for the listeners.
+					callbackListeners = false
+				}
+			}
+		}
+	} else {
+		scopedLog.Debug("Attempt to upsert invalid IP into ipcache layer")
+		return false
+	}
+
+	scopedLog.Debug("Upserting IP into ipcache layer")
+
+	// Update both maps.
+	ipc.ipToIdentityCache[ip] = newIdentity
+	// Delete the old identity, if any.
+	if found {
+		delete(ipc.identityToIPCache[cachedIdentity.ID], ip)
+		if len(ipc.identityToIPCache[cachedIdentity.ID]) == 0 {
+			delete(ipc.identityToIPCache, cachedIdentity.ID)
+		}
+	}
+	if _, ok := ipc.identityToIPCache[newIdentity.ID]; !ok {
+		ipc.identityToIPCache[newIdentity.ID] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[newIdentity.ID][ip] = struct{}{}
+
+	if callbackListeners {
+		for _, listener := range ipc.listeners {
+			listener.OnIPIdentityCacheChange(Upsert, *cidr, hostIP, oldIdentity, newIdentity.ID)
+		}
 	}
 
 	return true
 }
 
 // deleteLocked removes removes the provided IP-to-security-identity mapping
 // from ipc with the assumption that the IPCache's mutex is held.
-func (ipc *IPCache) deleteLocked(IP string) {
-	log.WithFields(logrus.Fields{
-		logfields.IPAddr: IP,
-	}).Debug("Removing from ipcache layer")
+func (ipc *IPCache) deleteLocked(ip string) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.IPAddr: ip,
+	})
 
-	identity, found := ipc.ipToIdentityCache[IP]
-	if found {
-		delete(ipc.ipToIdentityCache, IP)
-		delete(ipc.identityToIPCache[identity.ID], IP)
-		if len(ipc.identityToIPCache[identity.ID]) == 0 {
-			delete(ipc.identityToIPCache, identity.ID)
-		}
+	cachedIdentity, found := ipc.ipToIdentityCache[ip]
+	if !found {
+		scopedLog.Debug("Attempt to remove non-existing IP from ipcache layer")
+		return
+	}
+
+	var cidr *net.IPNet
+	cacheModification := Delete
+	var oldIdentity *identity.NumericIdentity
+	newIdentity := cachedIdentity
+	callbackListeners := true
 
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
 		// Remove a reference for the prefix length if this is a CIDR.
-		if _, cidr, err := net.ParseCIDR(IP); err == nil {
-			pl, bits := cidr.Mask.Size()
-			switch bits {
-			case net.IPv6len * 8:
-				unrefPrefixLength(ipc.v6PrefixLengths, pl)
-			case net.IPv4len * 8:
-				unrefPrefixLength(ipc.v4PrefixLengths, pl)
+		pl, bits := cidr.Mask.Size()
+		switch bits {
+		case net.IPv6len * 8:
+			unrefPrefixLength(ipc.v6PrefixLengths, pl)
+		case net.IPv4len * 8:
+			unrefPrefixLength(ipc.v4PrefixLengths, pl)
+		}
+
+		// Check whether the deleted CIDR was shadowed by an endpoint IP. In
+		// this case, skip calling back the listeners since they don't know
+		// about its mapping.
+		if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+			scopedLog.Info("Deleting CIDR shadowed by endpoint IP")
```

Comment:
Why does this need a info message?

Reply:
For some reason, the shadowing-related log messages were at info level. I just wanted to be consistent. I agree this should be debug.
Changed all of them.